File: backend/app/simulation/engine.py
"""Monte Carlo simulation engine for F1 Fantasy point predictions (2026 rules)."""

import numpy as np
from dataclasses import dataclass, field
from .scoring import (
    score_qualifying_driver,
    score_race_position,
    score_sprint_position,
    score_constructor_qualifying_progression,
    FASTEST_LAP_PTS,
    SPRINT_FASTEST_LAP_PTS,
    DRIVER_OF_THE_DAY_PTS,
    RACE_DNF_PENALTY,
    SPRINT_DNF_PENALTY,
)


@dataclass
class DriverParams:
    id: int
    code: str
    constructor_ref: str
    qpace_mean: float
    qpace_std: float
    dnf_probability: float
    fl_probability: float
    avg_positions_gained: float
    grid_penalty: int = 0


@dataclass
class ConstructorParams:
    id: int
    ref_id: str
    driver_ids: list[int] = field(default_factory=list)
    expected_pitstop_pts: float = 4.0


@dataclass
class SimResult:
    asset_type: str
    asset_id: int
    mean: float
    median: float
    std: float
    p10: float
    p90: float


def simulate_race_weekend(
    drivers: list[DriverParams],
    constructors: list[ConstructorParams],
    overtake_difficulty: float = 0.5,
    is_sprint: bool = False,
    n_simulations: int = 10000,
) -> list[SimResult]:
    n_drivers = len(drivers)
    rng = np.random.default_rng()

    driver_points = {d.id: np.zeros(n_simulations) for d in drivers}
    constructor_points = {c.id: np.zeros(n_simulations) for c in constructors}

    # Build lookup maps
    constructor_driver_map = {}
    for c in constructors:
        for did in c.driver_ids:
            constructor_driver_map[did] = c.id

    # Pre-generate random numbers
    quali_samples = rng.normal(
        loc=[d.qpace_mean for d in drivers],
        scale=[d.qpace_std for d in drivers],
        size=(n_simulations, n_drivers),
    )
    race_noise = rng.normal(
        loc=0, scale=[d.qpace_std * 1.2 for d in drivers],
        size=(n_simulations, n_drivers),
    )
    dnf_rolls = rng.random(size=(n_simulations, n_drivers))
    fl_rolls = rng.random(size=n_simulations)
    dotd_rolls = rng.random(size=n_simulations)

    overtake_ease = 1.0 - overtake_difficulty

    for sim in range(n_simulations):
        # === QUALIFYING ===
        raw_quali = quali_samples[sim]
        quali_order = np.argsort(raw_quali)
        quali_positions = np.empty(n_drivers, dtype=int)
        for rank, idx in enumerate(quali_order):
            quali_positions[idx] = rank + 1

        # === RACE ===
        race_positions = np.empty(n_drivers, dtype=int)
        is_dnf = np.zeros(n_drivers, dtype=bool)
        grid_positions = np.empty(n_drivers, dtype=int)

        for i, d in enumerate(drivers):
            # Grid position = qualifying + penalty (but quali pts use quali pos, not grid)
            grid_pos = quali_positions[i] + d.grid_penalty
            grid_pos = max(1, min(n_drivers, grid_pos))
            grid_positions[i] = grid_pos

            if dnf_rolls[sim, i] < d.dnf_probability:
                is_dnf[i] = True
                continue

            # Race pace: base + noise + overtake adjustment
            rpace = d.qpace_mean - d.avg_positions_gained * overtake_ease
            sampled = rpace + race_noise[sim, i]
            sampled = max(1, min(n_drivers, sampled))
            race_positions[i] = int(round(sampled))

        # Resolve race positions for non-DNF drivers
        active_indices = np.where(~is_dnf)[0]
        if len(active_indices) > 0:
            active_raw = race_positions[active_indices]
            active_order = np.argsort(active_raw)
            for rank, order_idx in enumerate(active_order):
                race_positions[active_indices[order_idx]] = rank + 1

        # === FASTEST LAP ===
        fl_weights = np.array([d.fl_probability if not is_dnf[i] else 0 for i, d in enumerate(drivers)])
        total_w = fl_weights.sum()
        if total_w > 0:
            fl_weights /= total_w
            cumsum = np.cumsum(fl_weights)
            fl_idx = np.searchsorted(cumsum, fl_rolls[sim])
            fl_idx = min(fl_idx, n_drivers - 1)
        else:
            fl_idx = 0

        # === DRIVER OF THE DAY (race only, not sprint) ===
        # DotD tends to go to drivers who had impressive races (big position gains)
        dotd_idx = -1
        if not is_sprint:
            dotd_weights = np.zeros(n_drivers)
            for i in range(n_drivers):
                if is_dnf[i]:
                    continue
                pos_gain = int(quali_positions[i]) - int(race_positions[i])
                # Drivers who gained lots of positions or finished on podium more likely
                weight = max(0.1, pos_gain + 2.0)
                if race_positions[i] <= 3:
                    weight += 3.0
                dotd_weights[i] = weight
            total_dw = dotd_weights.sum()
            if total_dw > 0:
                dotd_weights /= total_dw
                dotd_cumsum = np.cumsum(dotd_weights)
                dotd_idx = np.searchsorted(dotd_cumsum, dotd_rolls[sim])
                dotd_idx = min(dotd_idx, n_drivers - 1)

        # === SCORE DRIVERS ===
        for i, d in enumerate(drivers):
            pts = 0.0
            q_pos = int(quali_positions[i])

            # Qualifying points
            pts += score_qualifying_driver(q_pos)

            if is_dnf[i]:
                pts += SPRINT_DNF_PENALTY if is_sprint else RACE_DNF_PENALTY
            else:
                r_pos = int(race_positions[i])

                # Race/sprint finish points
                if is_sprint:
                    pts += score_sprint_position(r_pos)
                else:
                    pts += score_race_position(r_pos)

                # Positions gained/lost (quali position vs finish, per official rules)
                pos_change = q_pos - r_pos
                pts += pos_change

                # Overtakes: simulate as fraction of positions gained through actual on-track passes
                # Not all position gains are overtakes (some are from DNFs ahead, pit strategy)
                if pos_change > 0:
                    # Estimate ~60-80% of positions gained are real overtakes, scaled by track
                    overtake_ratio = 0.5 + overtake_ease * 0.3
                    estimated_overtakes = max(0, int(round(pos_change * overtake_ratio)))
                    pts += estimated_overtakes

                # Fastest lap
                if i == fl_idx:
                    pts += SPRINT_FASTEST_LAP_PTS if is_sprint else FASTEST_LAP_PTS

                # Driver of the Day (race only)
                if not is_sprint and i == dotd_idx:
                    pts += DRIVER_OF_THE_DAY_PTS

            driver_points[d.id][sim] = pts

        # === SCORE CONSTRUCTORS ===
        for c in constructors:
            c_pts = 0.0
            d_indices = [idx for idx, d in enumerate(drivers) if d.id in c.driver_ids]

            if len(d_indices) < 2:
                continue

            d0, d1 = d_indices[0], d_indices[1]
            q0 = int(quali_positions[d0])
            q1 = int(quali_positions[d1])

            # Constructor gets both drivers' qualifying points
            c_pts += score_qualifying_driver(q0) + score_qualifying_driver(q1)
            # Qualifying progression bonus
            c_pts += score_constructor_qualifying_progression(q0, q1)

            for di in [d0, d1]:
                if is_dnf[di]:
                    c_pts += SPRINT_DNF_PENALTY if is_sprint else RACE_DNF_PENALTY
                else:
                    r_pos = int(race_positions[di])

                    # Race/sprint finish
                    if is_sprint:
                        c_pts += score_sprint_position(r_pos)
                    else:
                        c_pts += score_race_position(r_pos)

                    # Positions gained/lost
                    pos_change = int(quali_positions[di]) - r_pos
                    c_pts += pos_change

                    # Overtakes for constructor
                    if pos_change > 0:
                        overtake_ratio = 0.5 + overtake_ease * 0.3
                        estimated_overtakes = max(0, int(round(pos_change * overtake_ratio)))
                        c_pts += estimated_overtakes

            # Pitstop points (race only)
            if not is_sprint:
                c_pts += c.expected_pitstop_pts

            constructor_points[c.id][sim] = c_pts

    # === COMPUTE RESULTS ===
    results = []
    for d in drivers:
        pts = driver_points[d.id]
        results.append(SimResult(
            asset_type="driver",
            asset_id=d.id,
            mean=float(np.mean(pts)),
            median=float(np.median(pts)),
            std=float(np.std(pts)),
            p10=float(np.percentile(pts, 10)),
            p90=float(np.percentile(pts, 90)),
        ))
    for c in constructors:
        pts = constructor_points[c.id]
        results.append(SimResult(
            asset_type="constructor",
            asset_id=c.id,
            mean=float(np.mean(pts)),
            median=float(np.median(pts)),
            std=float(np.std(pts)),
            p10=float(np.percentile(pts, 10)),
            p90=float(np.percentile(pts, 90)),
        ))

    return results

File: frontend/src/lib/simulation/engine.ts
/**
 * Monte Carlo simulation engine for F1 Fantasy point predictions (2026 rules).
 *
 * Ported from backend/app/simulation/engine.py — algorithmic parity
 * maintained via statistical tolerance tests in engine.golden.test.ts.
 *
 * Does NOT produce numpy-bit-identical output (Mulberry32 vs PCG64); relies
 * on n_simulations >= 10k for convergence to the same distribution.
 */
import {
  mulberry32,
  boxMullerNormal,
  sampleUniformArray,
} from "./rng";
import {
  scoreQualifyingDriver,
  scoreSprintQualifyingDriver,
  scoreRacePosition,
  scoreSprintPosition,
  scorePositionsChanged,
  scoreConstructorQualifyingProgression,
  FASTEST_LAP_PTS,
  DRIVER_OF_THE_DAY_PTS,
  RACE_DNF_PENALTY,
  SPRINT_DNF_PENALTY,
  BEAT_TEAMMATE_PTS,
  OVERTAKE_PTS,
} from "./scoring";

// ---------------------------------------------------------------------------
// Public interfaces (ported from Python @dataclass declarations)
// ---------------------------------------------------------------------------
export interface DriverParams {
  id: number;
  code: string;
  constructor_ref: string;
  qpace_mean: number;
  qpace_std: number;
  dnf_probability: number;
  fl_probability: number;
  avg_positions_gained: number;
  grid_penalty?: number;
}

export interface ConstructorParams {
  id: number;
  ref_id: string;
  driver_ids: number[];
  expected_pitstop_pts?: number;
  car_pace_std?: number;
}

export interface CircuitTraits {
  overtake_difficulty?: number;
  high_speed?: number;
  street_circuit?: boolean;
  altitude?: number;
  avg_degradation?: number;
}

export interface WeatherConfig {
  is_wet?: boolean;
  quali_std_multiplier?: number;
  race_noise_multiplier?: number;
  dnf_multiplier?: number;
}

export interface SimResult {
  asset_type: "driver" | "constructor";
  asset_id: number;
  mean: number;
  median: number;
  std: number;
  p10: number;
  p90: number;
}

export interface SimulateInput {
  drivers: DriverParams[];
  constructors: ConstructorParams[];
  circuit?: CircuitTraits;
  isSprint?: boolean;
  nSimulations?: number;
  weather?: WeatherConfig;
  seed: number;
}

// ---------------------------------------------------------------------------
// numpy-equivalent helpers (argsort, mean, median, std, percentile)
// Kept in this file (not rng.ts) because the semantics are tied to the
// engine's expected numpy defaults (population std, linear-interp percentile).
// ---------------------------------------------------------------------------

function argsort(arr: number[]): number[] {
  const indexed = arr.map((v, i): [number, number] => [v, i]);
  // Stable sort by value; ties keep original order (matches numpy.argsort kind='quicksort'
  // closely enough for the engine's purposes — ties in raw_quali are astronomically rare
  // for continuous normals).
  indexed.sort((a, b) => a[0] - b[0]);
  return indexed.map((pair) => pair[1]);
}

function mean(arr: number[]): number {
  if (arr.length === 0) return 0;
  let s = 0;
  for (let i = 0; i < arr.length; i++) s += arr[i];
  return s / arr.length;
}

/** numpy default: population std (ddof=0). */
function stdPopulation(arr: number[]): number {
  if (arr.length === 0) return 0;
  const m = mean(arr);
  let sq = 0;
  for (let i = 0; i < arr.length; i++) {
    const d = arr[i] - m;
    sq += d * d;
  }
  return Math.sqrt(sq / arr.length);
}

function median(arr: number[]): number {
  if (arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = sorted.length >> 1;
  return sorted.length % 2 === 0
    ? (sorted[mid - 1] + sorted[mid]) / 2
    : sorted[mid];
}

/**
 * numpy.percentile with method="linear" (the default in numpy >= 1.22).
 * Index = (p/100) * (n-1); linear interp between floor(idx) and ceil(idx).
 */
function percentile(arr: number[], p: number): number {
  if (arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  const frac = idx - lo;
  return sorted[lo] * (1 - frac) + sorted[hi] * frac;
}

// ---------------------------------------------------------------------------
// simulateRaceWeekend — direct line-for-line port of engine.py body.
// ---------------------------------------------------------------------------
export function simulateRaceWeekend(input: SimulateInput): SimResult[] {
  const drivers = input.drivers;
  const constructors = input.constructors;
  const circuit: Required<CircuitTraits> = {
    overtake_difficulty: input.circuit?.overtake_difficulty ?? 0.5,
    high_speed: input.circuit?.high_speed ?? 0.5,
    street_circuit: input.circuit?.street_circuit ?? false,
    altitude: input.circuit?.altitude ?? 0,
    avg_degradation: input.circuit?.avg_degradation ?? 0.5,
  };
  const weather: Required<WeatherConfig> = {
    is_wet: input.weather?.is_wet ?? false,
    quali_std_multiplier: input.weather?.quali_std_multiplier ?? 1.5,
    race_noise_multiplier: input.weather?.race_noise_multiplier ?? 1.4,
    dnf_multiplier: input.weather?.dnf_multiplier ?? 1.3,
  };
  const isSprint = input.isSprint ?? false;
  const nSimulations = input.nSimulations ?? 10000;
  const nDrivers = drivers.length;

  const rng = mulberry32(input.seed);

  const driverPoints = new Map<number, number[]>();
  for (const d of drivers) driverPoints.set(d.id, new Array(nSimulations).fill(0));
  const constructorPoints = new Map<number, number[]>();
  for (const c of constructors) constructorPoints.set(c.id, new Array(nSimulations).fill(0));

  // Build constructor_ref -> driver-index list (insertion-ordered, like Python dict)
  const constructorRefToIndices = new Map<string, number[]>();
  for (let i = 0; i < drivers.length; i++) {
    const ref = drivers[i].constructor_ref;
    if (!constructorRefToIndices.has(ref)) constructorRefToIndices.set(ref, []);
    constructorRefToIndices.get(ref)!.push(i);
  }

  // Teammate map
  const teammateMap = new Map<number, number>();
  for (const [, indices] of constructorRefToIndices) {
    if (indices.length === 2) {
      teammateMap.set(indices[0], indices[1]);
      teammateMap.set(indices[1], indices[0]);
    }
  }

  // Car pace std lookup
  const carStdMap = new Map<string, number>();
  for (const c of constructors) carStdMap.set(c.ref_id, c.car_pace_std ?? 1.5);

  // Pre-generate car offsets per constructor per sim (iteration order matches Python)
  const carOffsets = new Map<string, number[]>();
  for (const [refId] of constructorRefToIndices) {
    const std = carStdMap.get(refId) ?? 1.5;
    const arr = new Array<number>(nSimulations);
    for (let s = 0; s < nSimulations; s++) arr[s] = boxMullerNormal(rng, 0, std);
    carOffsets.set(refId, arr);
  }

  // Weather multipliers
  const qStdMult = weather.is_wet ? weather.quali_std_multiplier : 1.0;
  const rNoiseMult = weather.is_wet ? weather.race_noise_multiplier : 1.0;
  const dnfMult = weather.is_wet ? weather.dnf_multiplier : 1.0;

  // Pre-generate quali samples (nSimulations × nDrivers), row-major
  const qualiSamples: number[][] = new Array(nSimulations);
  for (let s = 0; s < nSimulations; s++) {
    const row = new Array<number>(nDrivers);
    for (let j = 0; j < nDrivers; j++) {
      row[j] = boxMullerNormal(
        rng,
        drivers[j].qpace_mean,
        drivers[j].qpace_std * qStdMult
      );
    }
    qualiSamples[s] = row;
  }

  // race_noise (nSimulations × nDrivers)
  const raceNoise: number[][] = new Array(nSimulations);
  for (let s = 0; s < nSimulations; s++) {
    const row = new Array<number>(nDrivers);
    for (let j = 0; j < nDrivers; j++) {
      row[j] = boxMullerNormal(rng, 0, drivers[j].qpace_std * 0.6 * rNoiseMult);
    }
    raceNoise[s] = row;
  }

  // dnf_rolls (nSimulations × nDrivers)
  const dnfRolls: number[][] = new Array(nSimulations);
  for (let s = 0; s < nSimulations; s++) {
    dnfRolls[s] = sampleUniformArray(rng, nDrivers);
  }

  const flRolls = sampleUniformArray(rng, nSimulations);
  const dotdRolls = sampleUniformArray(rng, nSimulations);

  const scProbability = circuit.street_circuit ? 0.45 : 0.55;
  const scRolls = sampleUniformArray(rng, nSimulations);

  const overtakeEase = 1.0 - circuit.overtake_difficulty;

  // ═════════════════════════════════════════════════════════════════════
  // Main simulation loop
  // ═════════════════════════════════════════════════════════════════════
  for (let sim = 0; sim < nSimulations; sim++) {
    // === QUALIFYING ===
    const rawQuali = qualiSamples[sim].slice();
    for (const [refId, indices] of constructorRefToIndices) {
      const offset = carOffsets.get(refId)![sim];
      for (const idx of indices) rawQuali[idx] += offset;
    }

    const qualiOrder = argsort(rawQuali);
    const qualiPositions = new Array<number>(nDrivers);
    for (let rank = 0; rank < qualiOrder.length; rank++) {
      qualiPositions[qualiOrder[rank]] = rank + 1;
    }

    // === GRID POSITIONS ===
    const gridPositions = new Array<number>(nDrivers);
    for (let i = 0; i < nDrivers; i++) {
      let gridPos = qualiPositions[i] + (drivers[i].grid_penalty ?? 0);
      gridPos = Math.max(1, Math.min(nDrivers, gridPos));
      gridPositions[i] = gridPos;
    }

    // === RACE ===
    const raceRaw = new Array<number>(nDrivers).fill(0);
    const racePositions = new Array<number>(nDrivers).fill(nDrivers + 1);
    const isDnf = new Array<boolean>(nDrivers).fill(false);

    for (let i = 0; i < nDrivers; i++) {
      const d = drivers[i];
      const effectiveDnf = d.dnf_probability * dnfMult;
      if (dnfRolls[sim][i] < effectiveDnf) {
        isDnf[i] = true;
        continue;
      }
      const deltaMean = -d.avg_positions_gained * overtakeEase;
      raceRaw[i] = gridPositions[i] + deltaMean + raceNoise[sim][i];
    }

    // Safety car — add mild noise to active drivers, then resolve positions
    const activeIndices: number[] = [];
    for (let i = 0; i < nDrivers; i++) if (!isDnf[i]) activeIndices.push(i);

    if (activeIndices.length > 0) {
      if (!isSprint && scRolls[sim] < scProbability) {
        for (const idx of activeIndices) {
          raceRaw[idx] += boxMullerNormal(rng, 0, 0.8);
        }
      }
      const activeRaw = activeIndices.map((idx) => raceRaw[idx]);
      const activeOrder = argsort(activeRaw);
      for (let rank = 0; rank < activeOrder.length; rank++) {
        racePositions[activeIndices[activeOrder[rank]]] = rank + 1;
      }
    }

    // === FASTEST LAP (race only, P1-P10) ===
    let flIdx = -1;
    if (!isSprint) {
      const flWeights = new Array<number>(nDrivers).fill(0);
      for (let i = 0; i < nDrivers; i++) {
        if (isDnf[i]) continue;
        const rPos = racePositions[i];
        if (rPos > 10) continue;
        const posFactor = rPos <= 5 ? 3.0 : 1.5;
        flWeights[i] = drivers[i].fl_probability * posFactor;
      }
      const totalW = flWeights.reduce((a, b) => a + b, 0);
      if (totalW > 0) {
        for (let i = 0; i < nDrivers; i++) flWeights[i] /= totalW;
        // Cumulative sum
        let acc = 0;
        const cumsum = flWeights.map((w) => (acc += w));
        // searchsorted(cumsum, fl_rolls[sim]) — leftmost index where cumsum[i] >= target
        const target = flRolls[sim];
        let picked = cumsum.findIndex((c) => c >= target);
        if (picked === -1) picked = nDrivers - 1;
        flIdx = Math.min(picked, nDrivers - 1);
        if (isDnf[flIdx] || racePositions[flIdx] > 10) flIdx = -1;
      }
    }

    // === DRIVER OF THE DAY (race only) ===
    let dotdIdx = -1;
    if (!isSprint) {
      const dotdWeights = new Array<number>(nDrivers).fill(0);
      for (let i = 0; i < nDrivers; i++) {
        if (isDnf[i]) continue;
        const rPos = racePositions[i];
        const posGain = gridPositions[i] - rPos;
        const moverWeight = Math.pow(Math.max(0, posGain), 1.5);
        let finishWeight = Math.max(0, (nDrivers - rPos) / nDrivers);
        if (rPos === 1) finishWeight += 2.0;
        else if (rPos <= 3) finishWeight += 1.0;
        dotdWeights[i] = moverWeight + finishWeight + 0.1;
      }
      const totalDw = dotdWeights.reduce((a, b) => a + b, 0);
      if (totalDw > 0) {
        for (let i = 0; i < nDrivers; i++) dotdWeights[i] /= totalDw;
        let acc = 0;
        const cumsum = dotdWeights.map((w) => (acc += w));
        const target = dotdRolls[sim];
        let picked = cumsum.findIndex((c) => c >= target);
        if (picked === -1) picked = nDrivers - 1;
        dotdIdx = Math.min(picked, nDrivers - 1);
      }
    }

    // === SCORE DRIVERS ===
    for (let i = 0; i < nDrivers; i++) {
      const d = drivers[i];
      let pts = 0;
      const qPos = qualiPositions[i];

      pts += isSprint
        ? scoreSprintQualifyingDriver(qPos)
        : scoreQualifyingDriver(qPos);

      const tmIdx = teammateMap.get(i);
      if (tmIdx !== undefined) {
        if (qPos < qualiPositions[tmIdx]) pts += BEAT_TEAMMATE_PTS;
      }

      if (isDnf[i]) {
        pts += isSprint ? SPRINT_DNF_PENALTY : RACE_DNF_PENALTY;
      } else {
        const rPos = racePositions[i];
        const gPos = gridPositions[i];

        pts += isSprint
          ? scoreSprintPosition(rPos)
          : scoreRacePosition(rPos);

        pts += scorePositionsChanged(gPos, rPos);

        const posChange = gPos - rPos;
        if (posChange > 0) {
          const overtakeRatio = 0.4 + overtakeEase * 0.4;
          const estimatedOvertakes = Math.max(
            0,
            Math.round(posChange * overtakeRatio)
          );
          pts += estimatedOvertakes * OVERTAKE_PTS;
        }

        if (i === flIdx) pts += FASTEST_LAP_PTS;
        if (!isSprint && i === dotdIdx) pts += DRIVER_OF_THE_DAY_PTS;

        if (tmIdx !== undefined) {
          if (isDnf[tmIdx]) pts += BEAT_TEAMMATE_PTS;
          else if (rPos < racePositions[tmIdx]) pts += BEAT_TEAMMATE_PTS;
        }
      }

      driverPoints.get(d.id)![sim] = pts;
    }

    // === SCORE CONSTRUCTORS ===
    for (const c of constructors) {
      let cPts = 0;
      const dIndices: number[] = [];
      for (let i = 0; i < drivers.length; i++) {
        if (c.driver_ids.includes(drivers[i].id)) dIndices.push(i);
      }
      if (dIndices.length < 2) continue;

      const [d0, d1] = [dIndices[0], dIndices[1]];
      const q0 = qualiPositions[d0];
      const q1 = qualiPositions[d1];

      if (isSprint) {
        cPts += scoreSprintQualifyingDriver(q0) + scoreSprintQualifyingDriver(q1);
      } else {
        cPts += scoreQualifyingDriver(q0) + scoreQualifyingDriver(q1);
      }
      cPts += scoreConstructorQualifyingProgression(q0, q1);

      for (const di of [d0, d1]) {
        if (isDnf[di]) {
          cPts += isSprint ? SPRINT_DNF_PENALTY : RACE_DNF_PENALTY;
        } else {
          const rPos = racePositions[di];
          const gPos = gridPositions[di];
          cPts += isSprint
            ? scoreSprintPosition(rPos)
            : scoreRacePosition(rPos);
          cPts += scorePositionsChanged(gPos, rPos);
          const posChange = gPos - rPos;
          if (posChange > 0) {
            const overtakeRatio = 0.4 + overtakeEase * 0.4;
            const estimatedOvertakes = Math.max(
              0,
              Math.round(posChange * overtakeRatio)
            );
            cPts += estimatedOvertakes * OVERTAKE_PTS;
          }
        }
      }

      if (!isSprint) cPts += c.expected_pitstop_pts ?? 4.0;

      constructorPoints.get(c.id)![sim] = cPts;
    }
  }

  // ═════════════════════════════════════════════════════════════════════
  // Aggregate results
  // ═════════════════════════════════════════════════════════════════════
  const results: SimResult[] = [];
  for (const d of drivers) {
    const pts = driverPoints.get(d.id)!;
    results.push({
      asset_type: "driver",
      asset_id: d.id,
      mean: mean(pts),
      median: median(pts),
      std: stdPopulation(pts),
      p10: percentile(pts, 10),
      p90: percentile(pts, 90),
    });
  }
  for (const c of constructors) {
    const pts = constructorPoints.get(c.id)!;
    results.push({
      asset_type: "constructor",
      asset_id: c.id,
      mean: mean(pts),
      median: median(pts),
      std: stdPopulation(pts),
      p10: percentile(pts, 10),
      p90: percentile(pts, 90),
    });
  }

  return results;
}
